packages/pwa-kit-react-sdk/src/ssr/universal/hooks/use-server-effect.tsx

```diff
@@ -0,0 +1,215 @@
+type EffectsResolver = () => Promise<{[x: string]: {}}>
+type Provider = ({ children, value }: { children: any; value: any; }) => JSX.Element
+type ServerEffect = (initialValue: Object, didUpdate: DidUpdateFn, source: any[]) => ServerEffectState
+type ServerEffectContext = {
+    /**
+    * ServerEffectProvider: The React provider you must wrap you application with in order to use
+    * the accompanied hook.
+    */
+    ServerEffectProvider: Provider; // Can we get more specifc here? provider type?
+    /**
+    * useServerEffect: A useEffect style hook that can be used client and server side to run async
+    * actions.
+    */
+    useServerEffect: ServerEffect;
+    /**
+    * resolveEffects: For use on the server side only, this async function will can be called to invoke
+    * all the recorded async "didUpdate" functions.
+    */
+    resolveEffects: EffectsResolver;
+}
+type ServerEffectContextValue = {
+    name: string;
+    data: object;
+    requests: DidUpdateFn[];
+    resolved: boolean;
+}
+
+/**
+ * This function is creating `useServerEffect`, it's only going to be used
+ * but the SDK, this isn't for external libs.
+ * 
+ * @param {*} context 
+ * @returns 
+ */
+const createServerEffect = (context) => {
+    const useServerEffect = (didUpdate: DidUpdateFn, source: any[], initial?: any) : ServerEffectState => {  
+        debugger  
+        const key = useUID()
+        const ctx: ServerEffectContextValue = useContext(context)
+
+        const getPropsArgs = useGetPropsArgs()
+        const [data, setData] = useState(ctx.data[key] || initial)
+        const [loading, setLoading] = useState(false)
+        const [error, setError] = useState(undefined)
+        const boundDidUpdate = didUpdate.bind(this, getPropsArgs)
+
+        const wrappedDidUpdate = isServer ? 
+            () => {} : 
+            async () => {
+                // Set loading state to "true"
+                setLoading(true)
+        
+                try {
+                    const data = await boundDidUpdate()
+                    setData(data)
+                } catch(e) {
+                    setLoading(false)
+                    setError(e)
+                }
+        
+                // Set loading state to false
+                setLoading(false)
+            }
+    
+        console.log('PUSHING REQUEST: ', !ctx.resolved, isServer)
+        if (!ctx.resolved && isServer) {
+            
+            ctx.requests.push(async () => {
+                const data = await boundDidUpdate()
+    
+                return {
+                    [key]: data
+                }
+            })
+        }
+    
+        // To avoid messing with React's hooks order lets make this a
+        // noop for the server.
+        useEffect(wrappedDidUpdate, source)
+    
+        return {
+            loading,
+            error,
+            data
+        }
+    }
+
+    return (...args) => (
+        useServerEffect.apply(this, args)
+    )
```

Comment:
```suggestion
    return useServerEffect
```
If I followed everything correctly, all functions wrapping this are arrow functions. Therefore, `this` is just `globalThis`, so this could just be `(...args) => useServerEffect(...args)`. Which, unless there's something weird React magic going on, should be equivalent to just returning `useServerEffect` directly.